libopeniscsiusr/idbm.c

```diff
@@ -0,0 +1,699 @@
+/* ^ MAX_KEYS, NAME_MAXVAL and VALUE_MAXVAL are copied from usr/idbm.h
+ * The RFC 3720 only said:
+ *	If not otherwise specified, the maximum length of a simple-value (not
+ *	its encoded representation) is 255 bytes, not including the delimiter
+ *	(comma or zero byte).
+ */
+
+#define OPTS_MAXVAL	8
+
+#define IDBM_HIDE	0    /* Hide parameter when print. */
+#define IDBM_SHOW	1    /* Show parameter when print. */
+#define IDBM_MASKED	2    /* Show "stars" instead of real value when print */
+
+#define ISCSI_BEGIN_REC	"# BEGIN RECORD "ISCSI_VERSION_STR
+#define ISCSI_END_REC	"# END RECORD"
+
+#ifndef LOCK_DIR
+#define LOCK_DIR		"/var/lock/iscsi"
+#endif
+#define LOCK_FILE		LOCK_DIR"/lock"
+#define LOCK_WRITE_FILE		LOCK_DIR"/lock.write"
+
+#define _rec_str(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_STR; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
+	if (strlen((char*)_org->_name)) \
+		strncpy((char*)_recs[_n].value, (char*)_org->_name, \
+			VALUE_MAXVAL); \
+	_recs[_n].data = &_org->_name; \
+	_recs[_n].data_len = sizeof(_org->_name); \
+	_recs[_n].visible = _show; \
+	_recs[_n].can_modify = _mod; \
+	_n++; \
+} while(0)
+
+#define _rec_uint8(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_UINT8; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
+	snprintf(_recs[_n].value, VALUE_MAXVAL, "%" PRIu8, _org->_name); \
+	_recs[_n].data = &_org->_name; \
+	_recs[_n].data_len = sizeof(_org->_name); \
+	_recs[_n].visible = _show; \
+	_recs[_n].can_modify = _mod; \
+	_n++; \
+} while (0)
+
+#define _rec_uint16(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_UINT16; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
+	snprintf(_recs[_n].value, VALUE_MAXVAL, "%" PRIu16, _org->_name); \
+	_recs[_n].data = &_org->_name; \
+	_recs[_n].data_len = sizeof(_org->_name); \
+	_recs[_n].visible = _show; \
+	_recs[_n].can_modify = _mod; \
+	_n++; \
+} while (0)
+
+#define _rec_uint32(_key, _recs, _org, _name, _show, _n, _mod) \
+do { \
+	_recs[_n].type = TYPE_UINT32; \
+	strncpy(_recs[_n].name, _key, NAME_MAXVAL); \
+	snprintf(_recs[_n].value, VALUE_MAXVAL, "%" PRIu16, _org->_name); \
+	_recs[_n].data = &_org->_name; \
+	_recs[_n].data_len = sizeof(_org->_name); \
+	_recs[_n].visible = _show; \
+	_recs[_n].can_modify = _mod; \
+	_n++; \
+} while (0)
+
+enum modify_mode {
+	_CANNOT_MODIFY,
+	_CAN_MODIFY,
+};
+
+struct idbm_rec {
+	int			type;
+	char			name[NAME_MAXVAL];
+	char			value[VALUE_MAXVAL];
+	void			*data;
+	int			data_len;
+	int			visible;
+	/*
+	 * TODO: make it a enum that can indicate wheter it also requires
+	 * a relogin to pick up if a session is running.
+	 */
+	enum modify_mode	can_modify;
+};
+
+int _idbm_lock(struct iscsi_context *ctx)
+{
+	int fd, i, ret;
+	struct idbm *db = NULL;
+	char strerr_buff[_STRERR_BUFF_LEN];
+
+	assert(ctx != NULL);
+
+	db = ctx->db;
+
+	if (db->refs > 0) {
+		db->refs++;
+		return 0;
+	}
+
+	if (access(LOCK_DIR, F_OK) != 0) {
```

Comment:
Race condition between checking the existence of creating it if it's missing.

Reply:
That's what old code doing and I guess we have to take that race.

Reply:
@tasleson please create a separate issue on that directory check/create race? thank you

Reply:
@gonzoleeman As requested: https://github.com/open-iscsi/open-iscsi/issues/95